src/main/java/com/exasol/adapter/document/files/AdlsRemoteFileFinder.java

```diff
@@ -0,0 +1,94 @@
+import com.exasol.adapter.document.iterators.CloseableIteratorWrapper;
+import com.exasol.adapter.document.iterators.FilteringIterator;
+import com.exasol.adapter.document.iterators.TransformingIterator;
+
+
+/**
+ * File finder for files on Azure Data Lake Storage Gen 2.
+ */
+public class AdlsRemoteFileFinder implements RemoteFileFinder {
+    private final StringFilter filePattern;
+    private final DataLakeFileSystemClient dlFileSystemClient;
+
+    /**
+     * Create a new instance of {@link AdlsRemoteFileFinder}.
+     *
+     * @param filePattern          pattern to search for
+     * @param connectionProperties connection information
+     */
+    //the c'tor sets up a Data Lake Storage Container/File System client and saves the file pattern for internal operations.
+    public AdlsRemoteFileFinder(final StringFilter filePattern, final AdlsConnectionProperties connectionProperties) {
+        //"storage" in GCP
+        final DataLakeServiceClient datalakeServiceClient = buildAdlsClient(connectionProperties);
+        //"container" is the equivalent to the "bucket" in GCP
+        this.dlFileSystemClient = datalakeServiceClient.getFileSystemClient(connectionProperties.getAdlsContainerName());
+        this.filePattern = filePattern;
+    }
+
+    private DataLakeServiceClient buildAdlsClient(final AdlsConnectionProperties connectionProperties) {
+        StorageSharedKeyCredential sharedKeyCredential =
+                new StorageSharedKeyCredential(connectionProperties.getAdlsStorageAccountName(), connectionProperties.getAdlsStorageAccountKey());
+        var builder = new DataLakeServiceClientBuilder();
+        builder.credential(sharedKeyCredential);
+        builder.endpoint("https://" + connectionProperties.getAdlsStorageAccountName() + ".dfs.core.windows.net");
+        return builder.buildClient();
+    }
+
+    @SuppressWarnings("java:S2095") // executorServiceFactory is closed by CloseInjectIterator
+    @Override
+    public CloseableIterator<RemoteFile> loadFiles() {
+        //get the files info
+        final CloseableIterator<AdlsObjectDescription> objectKeys = getQuickFilteredObjectKeys();
+        //filter files again
+        final com.exasol.adapter.document.files.stringfilter.matcher.Matcher filePatternMatcher = this.filePattern
+                .getDirectoryIgnoringMatcher();
+        final FilteringIterator<AdlsObjectDescription> filteredObjectKeys = new FilteringIterator<>(objectKeys,
+                s3Object -> filePatternMatcher.matches(s3Object.getName()));
+        //load all the files
+        final ExecutorServiceFactory executorServiceFactory = new ExecutorServiceFactory();
+        final CloseableIterator<RemoteFile> loadedFiles = new TransformingIterator<>(filteredObjectKeys,
+                description -> this.loadFile(description, executorServiceFactory));
+        return new CloseInjectIterator<>(loadedFiles, executorServiceFactory::close);
```

Comment:
This might be dangerous. When closing the `CloseableIterator` the executor service will be shutdown and `RemoteFile.loadAsync()` will fail.
Would it be possible to create a single `ExecutorService` in the entry point, inject it where necessary, and close it when everything is finished?
No need to do this in this PR but please create a ticket to do it afterwards.

Reply:
This is how it's done in all document VS right now. Should be discussed with JB as well.